Archive/Wiedijk100Theorems/BuffonsNeedle.lean

```diff
@@ -0,0 +1,384 @@
+
+As in many references, the problem is split into two cases, `l ≤ d` (`buffon_short`), and `l ≥ d`
+(`buffon_long`). For both cases, we show that
+```
+ℙ[N] = (d * π) ⁻¹ *
+    ∫ θ in 0..π,
+    ∫ x in Set.Icc (-d / 2) (d / 2) ∩ Set.Icc (-θ.sin * l / 2) (θ.sin * l / 2), 1
+```
+In the short case `l ≤ d`, we show that `[-l * θ.sin/2, l * θ.sin/2] ⊆ [-d/2, d/2]`
+(`short_needle_inter_eq`), and therefore the inner integral simplifies to
+```
+∫ x in (-θ.sin * l / 2)..(θ.sin * l / 2), 1 = θ.sin * l
+```
+Which then concludes in the short case being `ℙ[N] = (2 * l) / (d * π)`.
+
+In the long case, `d ≥ l` (`buffon_long`), we show the outer integral simplifies to
+```
+∫ θ in 0..π, min d (θ.sin * l)
+```
+which can be expanded to
+```
+2 * (∫ θ in 0..(d / l).arcsin, min d (θ.sin * l) + ∫ θ in (d / l).arcsin..(π/2), min d (θ.sin * l))
+```
+We then show the two integrals equal their respective values `l - (l^2 - d^2).sqrt` and
+`(π / 2 - (d / l).arcsin) * d`. Then with some algebra we conclude
+```
+ℙ[N] = (2 * l) / (d * π) - 2 / (d * π) * ((l^2 - d^2).sqrt + d * (d / l).arcsin) + 1
+```
+
+## References
+
+* https://en.wikipedia.org/wiki/Buffon%27s_needle_problem
+* https://www.math.leidenuniv.nl/~hfinkeln/seminarium/stelling_van_Buffon.pdf
+* https://www.isa-afp.org/entries/Buffons_Needle.html
+
+-/
+
+open MeasureTheory (MeasureSpace IsProbabilityMeasure pdf.IsUniform Measure)
+open ProbabilityTheory Real
+
+lemma set_integral_toReal_ofReal_nonneg_ae {α : Type _} [MeasureSpace α] {s : Set α} {f : α → ℝ}
+    (hs : MeasurableSet s) (hf : ∀ᵐ x : α, x ∈ s → f x ≥ 0) :
+    ∫ (x : α) in s, ENNReal.toReal (ENNReal.ofReal (f x)) =
+    ∫ (x : α) in s, f x := by
+
+  have : ∀ᵐ x : α, x ∈ s → f x = ENNReal.toReal (ENNReal.ofReal (f x)) := by
+    simp_rw [eq_comm (a := f _), ENNReal.toReal_ofReal_eq_iff, hf]
+
+  rw [MeasureTheory.set_integral_congr_ae hs this]
+
+variable
+  /- Probability theory variables. -/
+  {Ω : Type*} [MeasureSpace Ω] [IsProbabilityMeasure (ℙ : Measure Ω)]
+
+  /- Buffon's needle variables. -/
+
+  /-
+    - `d > 0` is the distance between parallel lines.
+    - `l > 0` is the length of the needle.
+  -/
+  (d l : ℝ)
+  (hd : d > 0)
+  (hl : l > 0)
+
+  /-
+    `B = (X, Θ)` is the joint random variable for the x-position and angle of the needle. If a line
+    is at `x = 0`, by symmetry, `B` can be a uniform random variable over `[-d/2, d/2] × [0, π]`,
+    where `θ` is the angle of the needle relative to the y-axis.
+  -/
+  (B : Ω → ℝ × ℝ)
+  (hBₘ : Measurable B)
+
+  /- `B` is uniformly distributed on `[-d/2, d/2] × [0, π]`. -/
+  (hB : pdf.IsUniform B ((Set.Icc (-d / 2) (d / 2)) ×ˢ (Set.Icc 0 π)) ℙ)
+
+/--
+  Projection of a needle onto the x-axis. The needle's center is at
+  x-coordinate `x`, of length `l` and angle `θ`. Note, `θ` is measured
+  relative to the y-axis, that is, a vertical needle has `θ = 0`.
+-/
+def needle_x_proj (x θ : ℝ) : Set ℝ := Set.Icc (x - θ.sin * l / 2) (x + θ.sin * l / 2)
```

Comment:
Can you use lowerCamelCase for terms of types (See [here](scoped[MeasureTheory]))

Reply:
This link doesn't work form me

Reply:
My bad, this is the link: https://github.com/leanprover-community/mathlib4/wiki/Lean-4-survival-guide-for-Lean-3-users#naming-conventions

Reply:
So I've read through this, do you have any recommendations on handling random variables (since they are typically just upper cased)? Should I just stick to the style guideline?

Reply:
In this case I would stick with the guidelines. OTOH if the random variable is a local variable within a theorem or a definition, then you can name it as you'd like (typically X)